contracts/Remittance.sol

```diff
@@ -1,16 +1,137 @@
 //SPDX-License-Identifier: MIT
 pragma solidity >= 0.6.0 < 0.7.0;
 
-import "../node_modules/@openzeppelin/contracts/access/Ownable.sol";
-import "../node_modules/@openzeppelin/contracts/utils/Pausable.sol";
+import "./access/Ownable.sol";
+import "./utils/Pausable.sol";
 
-import {SafeMath} from "../node_modules/@openzeppelin/contracts/math/SafeMath.sol";
+import {SafeMath} from "./math/SafeMath.sol";
 
 contract Remittance is Ownable, Pausable {
 
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
+    mapping(bytes32 => RemittanceInstance) public remittances;
 
+    // The 'valid' element provides a flag for checking a bonafide hash key
+    struct RemittanceInstance {
+        address funder;
+        address broker;
+        uint256 fundsOwed;
+        uint256 expiryDate;
+        bool valid;
+    }
+
+    event RemittanceCreated(
+        bytes32 indexed hash,
+        address indexed funder,
+        address indexed broker,
+        uint256 amount,
+        uint256 expiryDate
+    );
+
+    event RemittanceFundsReleased(
+        bytes32 indexed hash,
+        address indexed broker,
+        uint256 amount
+    );
+
+    event RemittanceFundsReclaimed(
+        bytes32 indexed hash,
+        address indexed funder,
+        uint256 amount
+    );
+
+    // The funder can create a remittance
+    // Hash is a sha256 hash of the broker address + recipient password
+    function create(bytes32 recipientPasswordHash, address broker, uint256 expiryDate)
+    external
+    payable
+    whenNotPaused
+    returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(msg.value > 0, "The amount must be greater than 0");
+
+        bytes32 rehash = keccak256ReHash(recipientPasswordHash, broker);
+
+        RemittanceInstance storage remittanceInstance = remittances[rehash];
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.broker = broker;
+        remittanceInstance.fundsOwed = msg.value;
```

Comment:
What happens if the funder calls the same function with the same `recipientPasswordHash` but different other parameters, including different `msg.value`?
Please answer before thinking of a fix.

Reply:
If the function was called a second time using the same hash and recipientPasswordHash
and broker address then the values would be overwritten.

I think some guard is required to check and then exit if existing values exist for that hash. Is that correct?

Reply:
The values would be overwritten, ... aaaaand?... this means thaaaat... ?

And yes the fix is to prevent overwriting.

Reply:
It means that, potentially, in a situation in which the funder or UI accidentally repeats the call/request then the contract would be funded in excess of the value stored in the remittance and thus permanently lost.

Reply:
Yep broken invariant. I burned your Ether my bad 🤷 
Do you know the story of Parity's multisig library?

Reply:
> Do you know the story of Parity's multisig library?

I have been reading about Parity's multisig library incident in more detail. I understand the nature of this attack from this explanation:

https://hackingdistributed.com/2017/07/22/deep-dive-parity-bug/